tests/tdastro/test_graph_state.py

```diff
@@ -402,6 +402,66 @@ def test_graph_to_from_file():
         state.save_to_file(file_path, overwrite=True)
 
 
+def test_graph_state_extract_parameters():
+    """Test that we can extract named parameters from a GraphState."""
+    state = GraphState()
+    state.set("a", "v1", 1.0)
+    state.set("a", "v2", 2.0)
+    state.set("a", "v3", 3.0)
+    state.set("b", "v1", 4.0)
+    state.set("c", "v2", 5.0)
+    state.set("c", "v3", 3.0)
+    state.set("d", "v4", 6.0)
+    state.set("e", "v3", 3.0)
+    state.set("e", "v5", 7.0)
+
+    # We can always access a parameter by its full name.
+    assert state["a.v1"] == 1.0
+    assert state["a.v2"] == 2.0
+
+    # With no filtering, we extract all the parameters.
+    results = state.extract_parameters()
+    assert len(results) == 9
+
+    # We can extract only certain parameters.
+    results = state.extract_parameters(["v2", "v3", "v4"])
+    assert len(results) == 6
+    assert results["a.v2"] == 2.0
+    assert results["a.v3"] == 3.0
+    assert results["c.v2"] == 5.0
+    assert results["c.v3"] == 3.0
+    assert results["e.v3"] == 3.0
+    assert results["d.v4"] == 6.0
+
+    # We can also provide a single parameter name.
+    results = state.extract_parameters("v2")
+    assert len(results) == 2
+    assert results["a.v2"] == 2.0
+    assert results["c.v2"] == 5.0
+
+    # We can extract from only certain nodes.
+    results = state.extract_parameters(nodes=["a", "c"])
+    assert len(results) == 5
+    assert results["a.v1"] == 1.0
+    assert results["a.v2"] == 2.0
+    assert results["a.v3"] == 3.0
+    assert results["c.v2"] == 5.0
+    assert results["c.v3"] == 3.0
+
+    # We can also provide a single node name.
+    results = state.extract_parameters(nodes="c")
+    assert len(results) == 2
+    assert results["c.v2"] == 5.0
+    assert results["c.v3"] == 3.0
+
+    # We can extract a set of parameters from a set of nodes.
+    results = state.extract_parameters(nodes=["a", "c", "e"], params=["v1", "v2"])
+    assert len(results) == 3
+    assert results["a.v1"] == 1.0
+    assert results["a.v2"] == 2.0
+    assert results["c.v2"] == 5.0
+
```

Comment:
I'm thinking something like `extract_parameters(['a.v1','c.v2','v3'])` which will return {'a.v1: 1, 'c.v2': 5, 'a.v3': 3, 'c.v3': 3, 'e.v3': 3}. Or `extract_parameters(['a.v1','v5'])` which will return {'a.v1': 1, 'v5': 7} or {'a.v1': 1, 'e.v5': 7} (either is fine but maybe the former is slightly preferred).
For example, I wanted to get a quick plot of the simulated Hubble Diagram using host redshift, so I'd want to do `extract_parameters(['host.redshift','x0','x1','c'])`, and I don't care (or don't know) what node `x0`,`x1` or `c` is from.

Reply:
That behavior was the point of the `collapse` logic. It allowed returning a single value for a parameter using just the parameter name.

In your case of `extract_parameters(['host.redshift','x0','x1','c'])` how do you want to handle duplicates? If we find instances of `x0` in two different nodes we could:
1. Throw an error [initial behavior]
2. Collapse them if they are identical
    2a. Throw an error if they are not identical
    2b. Return both in expanded form if they are not identical [second round behavior]
3. Return all values [current behavior]
4. Return an arbitrary one of the `x0` values. [Not recommended as the behavior becomes undefined].

More concretely if someone where to call `extract_parameters(['ra', 'dec', 't0'])` where the host and source have two different values of `ra`, how would you like the code to behave?

Reply:
I think we want to output duplicate parameter names with their node names (say `a.x0`, `b.x0`), regardless of whether they are identical in values.
If someone calls `extract_parameters(['ra','dec','t0'])`, we'll return {'host.ra','source.ra','anything_else.ra', ...}, even some of them may be identical in their values.

Reply:
I wanted to add that I think `extract_parameters(node=[],parameters=[])` is still useful. Maybe parameters can be expended to include `extract_parameters(parameters=['host.redshift','x0','x1','c'])`?

Reply:
I think I've got it, but let me check.

In the case of a state:
```
a.v1 -> 1.0
a.v2 -> 2.0
a.v3 -> 3.0
b.v1 -> 4.0
b.v2 -> 5.0
b.v4 -> 6.0
c.v1 -> 7.0
```
and the query `extract_parameters(parameters=['a.v1', 'v2', 'v3', 'b.v4'])` you would want the function to return the dictionary:

```
{
    "a.v1": 1.0,
    "a.v2": 2.0,
    "b.v2": 5.0,
    "v3": 3.0,
    "b.v4": 6.0,
}
```

Specifically:
1) The full name are always preserved (even if the parameter does not exist in other nodes).
2) The shortened name is preserved if the parameter is unique.
3) The shortened name is transformed into multiple long names if the parameter is not unique.
Is that correct?

Reply:
I implemented those rules. I've dropped the `nodes` parameter for now, because it is not clear how to incorporate that as well. We can always extract all parameters with `to_dict`.